File: src/main/kotlin/acc/business/balance/Balance.java
package acc.business.balance;

import acc.business.Facade;
import acc.model.AbstrGroup;
import acc.model.AnalAcc;
import acc.model.Transaction;
import acc.util.AccException;
import acc.util.Global;

import java.time.LocalDate;
import java.time.Month;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;

public class Balance {

    public static final Balance instance = new Balance();

    private Map<AbstrGroup, BalanceItem> bitems;
    private BalanceItem root;

    public List<BalanceItem> createBalance(Month month) throws AccException {
        bitems = new TreeMap<>();
        root = new BalanceItem();
        Facade.INSTANCE.getBalanceAccounts()
                .forEach(ba -> insertToTree(Optional.of(ba)));
        sumTransactions(month);
        root.sum();
        return root.appendItemsTo(new ArrayList<>());
    }

    private BalanceItem insertToTree(Optional<AbstrGroup> insertedItemGroup) {
        if (!insertedItemGroup.isPresent()) {
            return root;
        }
        AbstrGroup iGroup = insertedItemGroup.get();
        BalanceItem parentItem = insertToTree(Optional.ofNullable(iGroup.getParent()));
        Map<String, BalanceItem> siblings = parentItem.getChildren();
        BalanceItem newItem = siblings.get(iGroup.getNumber());
        if (newItem == null) {
            newItem = new BalanceItem(iGroup);
            bitems.put(iGroup, newItem);
            siblings.put(iGroup.getNumber(), newItem);
        }
        return newItem;
    }

    void add(AnalAcc acc, boolean inMonth, boolean isInit, long amount) {
        if (!acc.getBalanced()) {
            return;
        }
        BalanceItem item = bitems.get(acc);
        if (acc.isActive()) {
            if (isInit) {
                item.addInitAssets(amount);
                item.addFinalAssets(amount);
            } else {
                item.addAssetsSum(amount);
                item.addFinalAssets(amount);
                if (inMonth) {
                    item.addAssets(amount);
                }
            }
        } else {
            if (isInit) {
                item.addInitLiabilities(amount);
                item.addFinalLiabilities(amount);
            } else {
                item.addLiabilitiesSum(amount);
                item.addFinalLiabilities(amount);
                if (inMonth) {
                    item.addLiabilities(amount);
                }
            }
        }
    }

    private void sumTransactions(Month month) throws AccException {
        for (Transaction trans : Facade.INSTANCE.getAllTransactions()) {
            LocalDate begin = LocalDate.of(Global.INSTANCE.getYear(), month, 1).minusDays(1);
            LocalDate end = begin.plusMonths(1);
            boolean isInit = !trans.getDate().isPresent();
            boolean inMonth = !isInit && trans.getDate().get().isAfter(begin)
                    && trans.getDate().get().isBefore(end);

            add(trans.getDal(), inMonth, isInit, -trans.getAmount());
            add(trans.getMaDati(), inMonth, isInit, +trans.getAmount());

        }

    }

}

File: src/main/kotlin/acc/business/balance/Balance.kt
package acc.business.balance

import acc.business.Facade
import acc.model.AbstrGroup
import acc.model.AnalAcc
import acc.util.AccException
import acc.Options
import java.time.LocalDate
import java.time.Month
import java.util.*

class Balance {

    private val bitems: MutableMap<AbstrGroup, BalanceItem> = TreeMap()
    private val root: BalanceItem = BalanceItem()

    @Throws(AccException::class)
    fun createBalance(month: Month): List<BalanceItem> {
        bitems.clear()
        Facade.balanceAccounts
                .forEach { ba -> insertToTree(ba) }
        sumTransactions(month)
        root.sum()
        return root.appendItemsTo(ArrayList())
    }

    private fun insertToTree(insertedItemGroup: AbstrGroup?): BalanceItem {
        if (insertedItemGroup == null) {
            return root
        }
        val parentItem = insertToTree(insertedItemGroup.parent)
        val siblings: MutableMap<String, BalanceItem> = parentItem.children
        var newItem: BalanceItem? = siblings[insertedItemGroup.number]
        if (newItem == null) {
            newItem = BalanceItem(insertedItemGroup)
            bitems[insertedItemGroup] = newItem
            siblings[insertedItemGroup.number] = newItem
        }
        return newItem
    }

    private fun addInit(acc: AnalAcc) {
        val item = bitems[acc]!!
      //  println(acc.initAmount)
        if (acc.isActive) {
            item.addInitAssets(acc.initAmount)
            item.addFinalAssets(acc.initAmount)

        } else {
            item.addInitLiabilities(acc.initAmount)
            item.addFinalLiabilities(acc.initAmount)

        }
    }

    private fun addTrans(acc: AnalAcc, inMonth: Boolean, amount: Long) {
        val item = bitems[acc]!!
        if (acc.isActive) {
            item.addAssetsSum(amount)
            item.addFinalAssets(amount)
            if (inMonth) {
                item.addAssets(amount)
            }
        } else
            item.addLiabilitiesSum(amount)
        item.addFinalLiabilities(amount)
        if (inMonth) {
            item.addLiabilities(amount)
        }

    }

    @Throws(AccException::class)
    private fun sumTransactions(month: Month) {
        Facade.balanceAccounts.forEach {
            addInit(it)
        }
        Facade.allTransactions.stream()
                .forEach { trans ->
                    val monthBegin = LocalDate.of(Options.year, month, 1)
                    val nextMontBegin = monthBegin.plusMonths(1)
                    val transDate = trans.doc.date
                    val inMonth =
                            !transDate.isBefore(monthBegin) and transDate.isBefore(nextMontBegin)
                    val maDati = trans.maDati
                    if (maDati.balanced)
                        addTrans(maDati, inMonth, +trans.amount)
                    if (trans.dal.balanced)
                        addTrans(trans.dal, inMonth, -trans.amount)
                }

    }

}